### PR TITLE
Victor/fixkernelconstraints

### PR DESCRIPTION
**Pull request by vpicheny**
This PR fixes 3 things:

- in `find_best_model_initialization` we now use the training loss so that it also works if we don't use any prior
- we "squeeze" the hyperparameters by an epsilon before searching for new ones. This avoids the following bug: say at the previous BO iteration, the optimal lengthscale reached its upper bound; if `find_best_model_initialization` doesn't find any better set of hyperparameters, you would re-assign the upper bound to the hyperparameters, then the untransformed value is +Inf and you get an error message.
- the previous sampling would return a single value for all lengthscales. They are now all different.

### CODE REVIEW

File: trieste/models/model_interfaces.py [hunk 1]
```diff
@@ -615,11 +614,19 @@ def randomize_model_hyperparameters(model: gpflow.models.GPModel) -> None:
...
         elif param.prior is not None:
             param.assign(param.prior.sample())
+
+
+def squeeze_hyperparameters(model: gpflow.models.GPModel, epsilon: float = 1e-8) -> None:
+    for param in model.trainable_parameters:
+        if isinstance(param.bijector, tfp.bijectors.Sigmoid):
+            squeezed_param = tf.math.minimum(param, param.bijector.high - epsilon)
+            squeezed_param = tf.math.maximum(squeezed_param, param.bijector.low + epsilon)
+            param.assign(squeezed_param)
```
Comment (henrymoss) [inline trieste/models/model_interfaces.py]: Perhaps add a little comment saying what this does/why it is needed? e.g. just use what you wrote in the PR description

File: trieste/models/model_interfaces.py [hunk 2]
```diff
@@ -615,11 +614,19 @@ def randomize_model_hyperparameters(model: gpflow.models.GPModel) -> None:
     for param in model.trainable_parameters:
         if isinstance(param.bijector, tfp.bijectors.Sigmoid):
             sample = tf.random.uniform(
-                [1],
+                param.bijector.low.shape,
```
Comment (henrymoss) [inline trieste/models/model_interfaces.py:617]: Good spot

File: trieste/models/model_interfaces.py (hunk 1 above)
Comment (henrymoss) [reply trieste/models/model_interfaces.py]: This might  also be unit-testable

Comment (henrymoss) [reply trieste/models/model_interfaces.py]: For example

```
@random_seed
def test_squeeze_hyperparameters_when_param_at_edge_of_bounds() -> None:
    kernel = gpflow.kernels.RBF(variance=1.0, lengthscales=[0.1 + 1e-5,0.5 - 1e-5])
    upper = tf.cast([0.5, 0.5], dtype=tf.float64)
    lower = upper / 5
    
    kernel.lengthscales = gpflow.Parameter(
        kernel.lengthscales, transform=tfp.bijectors.Sigmoid(low=lower, high=upper)
    )

    squeeze_hyperparameters(model,epsilon=1e-2)

    npt.assert_array_less(kernel.lengthscales, [0.1,0.5] - 1e-2)
    npt.assert_array_less(-kernel.lengthscales, -[0.1,0.5] + 1e-2)
```

File: trieste/models/model_interfaces.py (hunk 2 above)
Comment (henrymoss) [reply trieste/models/model_interfaces.py:617]: Also, could do with a unit test.

Comment (henrymoss) [reply trieste/models/model_interfaces.py:617]: Something like

```
@random_seed
def test_randomize_model_hyperparameters_samples_different_values_for_multi_dimensional_params(
) -> None:
    kernel = gpflow.kernels.RBF(variance=1.0, lengthscales=[0.2, 0.2])
    upper = tf.cast([10.0] * dim, dtype=tf.float64)
    lower = upper / 100
    kernel.lengthscales = gpflow.Parameter(
        kernel.lengthscales, transform=tfp.bijectors.Sigmoid(low=lower, high=upper)
    )

    randomize_model_hyperparameters(kernel)

    npt.assert_raises(AssertionError, npt.assert_allclose, kernel.lengthscales[0], kernel.lengthscales[1])


```

File: trieste/models/model_interfaces.py (hunk 1 above)
Comment (uri-granta) [reply trieste/models/model_interfaces.py]: (Yes, a docstring like for randomize_model_hyperparameters would be good.)

**Review by uri-granta — APPROVED** [review]
Looks good! Regarding tests, apart from @henrymoss's suggestion of testing the single-value-return bug fix and the squeeze parameters implementation, it might also be good to have explicit tests for the other two points of the ticket description:

1. check that `find_best_model_initialization` now handles params with no priors (see `test_find_best_model_initialization_improves_likelihood` and `test_find_best_model_initialization_only_changes_params_with_priors`)
2. check that `find_best_model_initialization` doesn't result in an +inf error if the optimal lengthscales has already reached its upper bound.

File: trieste/models/model_interfaces.py (hunk 1 above)
Comment (vpicheny) [reply trieste/models/model_interfaces.py]: thanks! done

File: trieste/models/model_interfaces.py (hunk 2 above)
Comment (vpicheny) [reply trieste/models/model_interfaces.py:617]: thanks! done
